synapseutils/migrate_functions.py

```diff
@@ -0,0 +1,736 @@
+    # just a limit on certain operations to put an upper bound on various
+    # batch operations so they are chunked. a function to make it easily mocked.
+    # don't anticipate needing to adjust this for any real activity
+    return 500
+
+
+def _ensure_tables(cursor):
+    # ensure we have the sqlite schema we need to be able to record and sort our
+    # entity file handle migration.
+
+    cursor.execute(
+        """
+            create table if not exists entities (
+                id text primary key,
+                parent_id text null,
+                type text not null,
+                status integer not null,
+                exception text null
+            );
+        """
+    )
+
+    cursor.execute(
+        """
+            create table if not exists file_entity_versions (
+                id text,
+                version integer,
+                from_file_handle_id text not null,
+                to_file_handle_id text not null,
+                primary key(id, version),
+                foreign key (id) references entities(id)
+            );
+        """
+    )
+
+    cursor.execute(
+        """
+            create table if not exists table_entity_files (
+                id text,
+                row integer,
+                column text,
+                from_file_handle_id text not null,
+                to_file_handle_id text not null,
+                primary key(id, row, column),
+                foreign key (id) references entities(id)
+            );
+        """
+    )
+
+
+def _wait_futures(conn, cursor, futures, return_when, continue_on_error):
+    completed, futures = concurrent.futures.wait(futures, return_when=return_when)
+    for completed_future in completed:
+        try:
+            entity_id, record_fn, mapping = completed_future.result()
+            record_fn(cursor, entity_id, mapping)
+
+            cursor.execute(
+                'update entities set status = ? where id = ?',
+                (_MigrationStatus.MIGRATED.value, entity_id)
+            )
+            conn.commit()
+
+        except _EntityMigrationError as ex:
+            logging.exception('Encountered error when migrating entity')
+
+            # for the purposes of recording and re-raise we're not interested in
+            # the _EntityMigrationError, just the underlying cause
+
+            cause = ex.__cause__
+            tb_str = ''.join(traceback.format_exception(type(cause), cause, cause.__traceback__))
+            cursor.execute(
+                'update entities set status = ?, exception = ? where id = ?',
+                (_MigrationStatus.MIGRATION_ERROR.value, tb_str, ex.entity_id)
+            )
+            conn.commit()
+
+            if not continue_on_error:
+                raise cause from None
+
+    conn.commit()
+    return futures
+
+
+def migrate(
+        syn,
+        entity,
+        storage_location_id,
+        version='new',
+        db_path=None,
+        continue_on_error=False,
+):
```

Comment:
If run from an interactive shell ask the user before proceeding with the migration.

Reply:
Migration now requires either interactive confirmation or using a force parameter.